### ISSUE TITLE
Slow animations

### ISSUE THREAD

**Issue by nikolaydevico** [heap/react-native-heap#235]
We have a pan animation in our application and it's extremely slow when react-native-heap library is installed. 
Is it possible to not just `ignore` but totally not handle some gesture elements? I didn't find anything it docs and HeapIgnore does not help to speed up animations. 

OS: iOS 14.2
React-native: 0.63.2

I can't give more info because I don't really understand the main problem and why this happens.

**Comment by jerryhjones** [comment]
Hey @nikolaydevico, I suspect this is caused by an iOS SDK bug that has recently been fixed, but has not been fully rolled out yet. Specifically, that bug was the result of over-traversing the view hierarchy during an event. In some cases where events are generated in rapid succession, such as certain animations and certain gestures, the Heap SDK could have a negative performance impact.

Do you possibly have a sample that reproduces this behavior that you could share with us? We'd love to confirm and reproduce the issue locally to validate that we've fully fixed it in the upcoming release.

**Comment by nikolaydevico** [comment]
Hello @jerryhjones , thank you for reply.
I'll prepare repo with example and description in coming days. 
Will let you know, thanks again.

**Comment by nikolaydevico** [comment]
Hey @jerryhjones, I've prepared repo and tested it https://github.com/nikolaydevico/stream-chat-react-native/tree/master/examples/SampleApp

Please follow instructions to install and run application on iOS and after that you need to:
- Select any user from the list
- Select any chat from the list
- Click on attachments icon on the bottom left corner of the screen and you should see attachments picker which could be swiped up. This attachment picker works almost ok with less than 10 files, but if you add more than 20 images to emulator/phone then it's almost impossible to use it. With 60+ images it totally freezes. Without heap library everything works ok.

Please let me know if you have any questions.

**Comment by sharplet** [comment]
@jerryhjones Any updates on that fix being rolled out? We've had to completely remove the Heap SDK from our app because scrolling performance made it unusable.

**Comment by airbugg** [comment]
@jerryhjones I was hoping for an update on this issue as well. 

Updating to the latest version (0.15.0) didn't help us resolve the issue (our react-native app still hangs when animations are involved).

**Comment by jerryhjones** [comment]
Hey folks - sorry for the slow reply on this. I want to give a little bit of an update.

Initially, we saw reports of slow animations and scrollviews in iOS that seemed to correspond in timing to these reports of React Native having similar issues. Because a great deal of our React Native library sits on top of our native iOS and Android libraries, we wrongly assumed that the root cause had been addressed when we fixed the bug in the underlying iOS library.

TL;DR There are some scenarios where user interaction causes Heap and React Native to interact in a way that results in poor performance. See below for a workaround.

**Why does this happen?**
When a user action takes place, Heap walks _up_ the view hierarchy to gather contextual information about the view that was tapped. In a previously release of the iOS library, this was unbounded, and could cause performance issues in complex view hierarchies - it now defaults to 30 parent views. The two items gathered about each view are a) its `class` b) its `accessibilityLabel`

However, React Native's `-[RCTView accessibilityLabel]` is [notoriously](https://github.com/facebook/react-native/pull/29801) [poor performing](https://github.com/facebook/react-native/pull/31222/), as it is possible for it to recursively walk the _entire_ view tree in worst case scenarios, while generating strings in a very inefficient manner.

**Workaround & Next Steps**
Ultimately, the native iOS ancestry data is unnecessary, as we also capture this information within at the React Native layer, which is the actual info we use on Heap events. We will be shipping updates to the iOS and React Native SDKs shortly to apply more sensible defaults that bypass iOS native hierarchy walking for React Native installs.

In the mean time, you can work around this issue by editing `RNHeapInit.m`, located in the `react-native-heap` development Pod.

Look for `HeapOptions *options = [[HeapOptions alloc] init];`, and add `options.hierarchyCaptureLimit = -1;` immediately after it. Your file should look like this:

```
#import "RNHeapInit.h"
#import <Foundation/Foundation.h>
#import "Heap.h"

@implementation RNHeapInit

+ (void)load {
    NSBundle *heapBundle = [NSBundle bundleWithPath:[[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:@"HeapSettings.bundle"]];
    NSString *heapPlistPath = [heapBundle pathForResource:@"HeapSettings" ofType:@"plist"];
    NSDictionary *heapPlistData = [NSDictionary dictionaryWithContentsOfFile:heapPlistPath];

    #ifdef DEBUG
        NSString *heapAppId = heapPlistData[@"HeapAppIdDev"];
        NSString *enableAutocapture = heapPlistData[@"HeapEnableAutocaptureDev"];
    #else
        NSString *heapAppId = heapPlistData[@"HeapAppIdProd"];
        NSString *enableAutocapture = heapPlistData[@"HeapEnableAutocaptureProd"];
    #endif

    if ([heapAppId length] > 0) {
        NSLog(@"Auto-initializing the Heap library with app ID %@ with native autocapture enabled=%@.", heapAppId, enableAutocapture);

        HeapOptions *options = [[HeapOptions alloc] init];
        // hierarchyCaptureLimit of -1 will prevent any hierarchy capture
        options.hierarchyCaptureLimit = -1;
        if (![enableAutocapture isEqualToString:@"true"]) {
            options.disableTouchAutocapture = YES;
        }

        [Heap initialize:heapAppId withOptions:options];
    } else {
        NSLog(@"Not auto-initializing Heap library.");
    }
}

@end

```

We're sorry that this has taken so long to address. We're taking this opportunity to improve our release processes, and are looking to add learnings to our performance metrics.

**Comment by bnickel** [comment]
This change has finally shipped as part of 0.17.0.

https://github.com/heap/react-native-heap/blob/3fa216b730d108a230dd08b085d958c800d01237/ios/RNHeapInit.m#L31-L34

**Comment by abriandev** [comment]
> This change has finally shipped as part of 0.17.0.
> 
> https://github.com/heap/react-native-heap/blob/3fa216b730d108a230dd08b085d958c800d01237/ios/RNHeapInit.m#L31-L34

@bnickel I am facing issue I run my app on android works fine and show autotrack logs on heap analytics but in ios It did'nt show autotrack logs on heapanalytics i followed the documentation but still not showing autotrack log on heap analysis can you tell me why this happens only on ios i tried on my ios device also still no response
